### PR TITLE
TSI-over-UART

### PR DESCRIPTION
**Pull request by jerryz123**
A PC host running the special `uart-tsi` fesvr implementation can communicate with a serial-adapter on a FPGA prototype or FPGA bringup-platform via UART.

Changes:
 - Allows forcing the SimUART DPI thing to put the UART on a pty
 - Fixes missing initialization of `is_loadmem` for testchip_tsi_t
 - Fixes bytes getting dropped out of the SimUART FIFO

Adds:
 - UART-to-TSI widget - intended for implementation on a FPGA board
 - `uart-tsi` FESVR implementation, tsi commands are sent out over a TTY
 
 Usage examples:
 
 Running a program
```
>  ./uart_tsi +tty=/dev/ttyUSB1 dhrystone.riscv
Attempting to open TTY at /dev/ttyUSB1
Constructed uart_tsi_t
Checking connection status with /dev/ttyUSB1
Connection succeeded
Microseconds for one run through Dhrystone: 452
Dhrystones per Second:                      2211
mcycle = 226186
minstret = 187526
Done, shutting down, flushing UART
```

Probing some register
```
> ./uart_tsi +tty=/dev/ttyUSB1 +init_read=0x10040 none
Attempting to open TTY at /dev/ttyUSB1
Constructed uart_tsi_t
Checking connection status with /dev/ttyUSB1
Connection succeeded
Reading 10040 ... got 517
```

Writing some register before running a program
```
> ./uart_tsi +tty=/dev/ttyUSB1 +init_write=0x88000000:0xdeadbeef test.riscv
Attempting to open TTY at /dev/ttyUSB1
Constructed uart_tsi_t
Checking connection status with /dev/ttyUSB1
Connection succeeded
Writing 88000000 with deadbeef
Done writing 88000000 with deadbeef
Hello world
Target reading from 0x88000000 - got deadbeef
Done, shutting down, flushing UART
```

### CODE REVIEW

File: src/main/resources/testchipip/csrc/testchip_tsi.h [hunk 1]
```diff
@@ -6,6 +6,12 @@
 #include <fesvr/tsi.h>
 #include <fesvr/htif.h>
 
+struct init_access_t {
+  uint64_t address;
+  uint32_t stdata;
```
Comment (harrisonliew) [inline src/main/resources/testchipip/csrc/testchip_tsi.h:11]: Why is data 32 bits only? Is it possible to support 8/16/32/64 bit data?

File: src/main/resources/testchipip/csrc/testchip_tsi.h [hunk 2]
```diff
@@ -32,6 +38,6 @@ class testchip_tsi_t : public tsi_t
 
   bool is_loadmem;
   bool write_hart0_msip;
-  std::vector<std::pair<uint64_t, uint32_t>> init_writes;
+  std::vector<init_access_t> init_accesses;
```
Comment (harrisonliew) [inline src/main/resources/testchipip/csrc/testchip_tsi.h:41]: Does this mean that we can have as many `+init_write`s and `+init_read`s as we want before the binary?

**Review by harrisonliew — CHANGES_REQUESTED** [review]

File: src/main/resources/testchipip/csrc/testchip_tsi.h (hunk 2 above)
Comment (jerryz123) [reply src/main/resources/testchipip/csrc/testchip_tsi.h:41]: Yes

File: src/main/resources/testchipip/csrc/testchip_tsi.h (hunk 1 above)
Comment (jerryz123) [reply src/main/resources/testchipip/csrc/testchip_tsi.h:11]: I'll leave this to future work.

**Review by harrisonliew — APPROVED** [review]

**Pull request merged**